### PR TITLE
Extend alternative module type loading to EventSetup

### LINKED ISSUE #435: Extend "alternative module loading" infrastructure to ESModules
I.e. this PR https://github.com/cms-sw/cmssw/pull/37731 (https://github.com/makortel/framework/issues/392). I think the corresponding place in EventSetup system is around
https://github.com/cms-sw/cmssw/blob/e7a6a4e6c3078e137a31bd44f89300f9d1f4861f/FWCore/Framework/src/EventSetupProviderMaker.cc#L95-L104
and 
https://github.com/cms-sw/cmssw/blob/e7a6a4e6c3078e137a31bd44f89300f9d1f4861f/FWCore/Framework/interface/ComponentFactory.h#L58-L61

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR extends the alternative module type loading, that was introduced in #37731 for EDModules, to EventSetup sources and modules. I'm not sure if this feature would be really useful for ESSources, because for GPUs the use case is really about transferring data from existing ESProducts on existing Records, but it was tiny bit simpler to allow the mechanism also for ESSources. For Loopers the mechanism is explicitly disabled.

Resolves https://github.com/makortel/framework/issues/435

#### PR validation:

Framework unit tests run.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: FWCore/Framework/interface/EventSetupProvider.h [hunk 1]
```diff
@@ -92,7 +93,8 @@ namespace edm {
           EventSetupProvider& precedingESProvider,
           std::set<ParameterSetIDHolder>& sharingCheckDone,
           std::map<EventSetupRecordKey, std::vector<ComponentDescription const*>>& referencedESProducers,
-          EventSetupsController& esController);
+          EventSetupsController& esController,
+          ModuleTypeResolverBase const* resolver);
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/EventSetupProvider.h:97]: I'd prefer having this moved to the first argument as I normally place mutable arguments at the end of function call argument lists.

File: FWCore/Framework/src/Factory.cc [hunk 2]
```diff
@@ -24,65 +25,13 @@ namespace edm {
...
-            //failed to find a plugin
-            return MakerPluginFactory::get()->create(modtype);
-          } catch (cms::Exception& iExcept) {
-            annotateExceptionAndRethrow(iExcept, p, modtype, resolver);
-          }
-        }
-        return MakerPluginFactory::get()->create(modtype);
-      };
-      std::unique_ptr<Maker> wm = make(resolver, modtype, p);
+      std::unique_ptr<Maker> wm = detail::resolveMaker<MakerPluginFactory>(modtype, resolver);
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Factory.cc:34]: nice sharing

File: FWCore/Framework/interface/EventSetupProvider.h (hunk 1 above)
Comment (makortel) [reply FWCore/Framework/interface/EventSetupProvider.h]: Changed the order. Also in another function where it seemed to make sense.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**